File: src/main/java/com/peasenet/settings/ColorSetting.java
/*
 * Copyright (c) 2022. Gavin Pease and contributors.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 *  of the Software, and to permit persons to whom the Software is furnished to do so, subject to the
 *  following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
 * OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

package com.peasenet.settings;

import com.peasenet.gavui.GuiCycle;
import com.peasenet.gavui.color.Color;
import com.peasenet.gavui.color.Colors;
import com.peasenet.gavui.util.GavUISettings;
import com.peasenet.main.Settings;
import net.minecraft.text.Text;

/**
 * @author gt3ch1
 * @version 7/1/2022
 * A setting that allows the user to change a color value.
 */
public class ColorSetting extends Setting {

    /**
     * The cycle element that allows the user to change the color value.
     */
    private final GuiCycle guiCycle;

    /**
     * The color value of the setting.
     */
    private Color color;

    private boolean isGavUi;

    /**
     * Creates a new color setting.
     *
     * @param name           - The name of the setting (ie, "foregroundColor").
     * @param translationKey - The translation key of the setting.
     */
    public ColorSetting(String name, String translationKey) {
        super(name, translationKey);
        if (name.contains("gavui")) {
            name = name.replace("gavui.", "");
            isGavUi = true;
            color = GavUISettings.getColor(name);
        } else color = Colors.WHITE;
        guiCycle = new GuiCycle(90, 10, Text.translatable(translationKey), Colors.COLORS.length);
        guiCycle.setBackground(color);
        String finalName = name;
        guiCycle.setCallback(() -> {
            color = Colors.COLORS[guiCycle.getCurrentIndex()];
            guiCycle.setBackground(color);
            if (isGavUi) {
                GavUISettings.add(finalName, color);
                GavUISettings.save();
            } else {
                Settings.add(finalName, color);
                Settings.save();
            }
            onClick();
        });
        guiCycle.setCurrentIndex(Colors.getColorIndex(color));
    }

    public ColorSetting(String translationKey) {
        this("none", translationKey);
    }

    /**
     * Sets the color and color index to the given value.
     *
     * @param index - The color index.
     */
    public void setColorIndex(int index) {
        guiCycle.setCurrentIndex(index);
        guiCycle.setBackground(Colors.COLORS[index]);
    }

    @Override
    public GuiCycle getGui() {
        return guiCycle;
    }

    /**
     * Gets the current color value.
     *
     * @return The current color value.
     */
    public Color getColor() {
        return color;
    }

    /**
     * Sets the background color of this setting.
     *
     * @param color - The color of the background.
     */
    public void setColor(Color color) {
        this.color = color;
        guiCycle.setBackground(color);
        guiCycle.setCurrentIndex(Colors.getColorIndex(color));
    }
}

File: src/main/kotlin/com/peasenet/settings/ColorSetting.kt
/*
 * MIT License
 *
 * Copyright (c) 2022-2025, Gavin C. Pease
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.peasenet.settings

import com.peasenet.gavui.GavUI
import com.peasenet.gavui.GuiBuilder
import com.peasenet.gavui.GuiCycle
import com.peasenet.gavui.color.Color
import com.peasenet.gavui.color.Colors
import com.peasenet.gavui.math.PointF

/**
 * A setting that allows the user to change a color value.
 * @author GT3CH1
 * @version 07-18-2023
 */
class ColorSetting(
    topLeft: PointF = PointF(0F, 0F),
    width: Float = 0F,
    height: Float = 10F,
    title: String = "",
    hoverable: Boolean = true,
    transparency: Float = -1f,
    symbol: Char = '\u0000',
    cycleIndex: Int = 0,
    cycleSize: Int = 0,
    color: Color = GavUI.backgroundColor(),
    override var callback: ((ColorSetting) -> Unit)? = null,
) : CallbackSetting<ColorSetting>(
    topLeft,
    width,
    height,
    title,
    hoverable = hoverable,
    transparency = transparency,
    symbol = symbol,
    cycleIndex = cycleIndex,
    cycleSize = cycleSize,
    color = color,
    callback = callback
) {
    override var gui: GuiCycle = GuiCycle(GuiBuilder())

    fun build(): ColorSetting {
        gui = GuiBuilder<GuiCycle>()
            .setWidth(width)
            .setHeight(height)
            .setTitle(title)
            .setCallback {
//                var index = it.currentIndex
//                index = (index + 1) % Colors.COLORS.size
//                it.currentIndex = index
                it.setBackground(Colors.COLORS[it.currentIndex])
                callback?.invoke(this)
            }
            .setHoverable(hoverable)
            .setCycleSize(Colors.COLORS.size)
            .setBackgroundColor(color)
            .setCurrentCycleIndex(Colors.COLORS.indexOf(color))
            .setTopLeft(topLeft)
            .buildCycle()
        return this
    }

    override var color: Color = color
        get() {
            return gui.backgroundColor
        }
        set(value) {
            gui.setBackground(value)
            field = value
        }
}

fun colorSetting(init: ColorSetting.() -> Unit): ColorSetting {
    val setting = ColorSetting()
    setting.init()
    return setting.build()
}